sequencer/src/request_response/recipient_source.rs

```
@@ -1,44 +1,70 @@
+
+#[derive(Clone)]
+pub struct RecipientSource<I: NodeImplementation<SeqTypes>, V: Versions> {
+    /// A copy of the consensus context
+    pub consensus: Consensus<I, V>,
+    /// A copy of the membership coordinator
+    pub memberships: EpochMembershipCoordinator<SeqTypes>,
+    /// The public key of the node
+    pub public_key: PubKey,
 }
 
 /// Implement the RecipientSourceTrait, which allows the request-response protocol to derive the
 /// intended recipients for a given request
 #[async_trait]
-impl RecipientSourceTrait<Request, PubKey> for RecipientSource {
-    async fn get_expected_responders(&self, request: &Request) -> Vec<PubKey> {
-        match request {
-            Request::Example => {
-                // Get the memberships
-                let memberships = self.memberships.read().await;
-
-                // Get everyone in the stake table
+impl<I: NodeImplementation<SeqTypes>, V: Versions> RecipientSourceTrait<Request, PubKey>
+    for RecipientSource<I, V>
+{
+    async fn get_expected_responders(&self, _request: &Request) -> Result<Vec<PubKey>> {
+        // Get the current epoch number
+        let epoch_number = self
+            .consensus
+            .consensus()
+            .read()
+            .await
+            .cur_epoch()
+            .unwrap_or(EpochNumber::genesis());
 
-                // NOTE: This is just an example request, therefore this is hardcoded to get members of epoch 0.
-                // When doing a real request, please choose the appropriate epoch to get members from.
-                memberships
-                    .stake_table(Some(EpochNumber::new(0)))
-                    .iter()
-                    .map(|entry| entry.stake_table_entry.stake_key)
-                    .collect()
+        // Attempt to get the membership for the current epoch
+        let membership = match self.memberships.wait_for_catchup(epoch_number).await {
+            Ok(membership) => membership,
```

@bfish713 this works - but is there a better function? If I call this on Epoch 0 or 1 it consistently gives me the warning and falls back to `membership_for_epoch`

---

Yeah I can do that. I asked Brendon what to use and he said either

---

I did this, but I think it makes more sense not to use `wait_for_catchup` at all (from what I understand, `membership_for_epoch` is the fallible version). Lmk what you think

---

yeah, it totally depends if you want to return an error and not wait

---

I think returning an error is fine here since we will just print out the error and re-request. Plus it will fall back to genesis anyway